File: typescript/src/coasty/webhooks.ts
/**
 * Webhook signature verification for `Coasty-Signature: t=<unix>,v1=<hex>`.
 *
 * Signed payload = `"<t>." + raw_body`;
 * `v1 = hex(HMAC_SHA256(webhook_secret, signed_payload))`.
 *
 * Verification uses `crypto.timingSafeEqual` (constant-time, with an
 * equal-length guard) and enforces a timestamp tolerance (default ±300s, the
 * documented replay window). Malformed input returns `false` — this function
 * never throws.
 */
import { createHmac, timingSafeEqual } from 'node:crypto';

export const DEFAULT_TOLERANCE_SECONDS = 300;

export interface VerifySignatureOptions {
  /** Allowed |now - t| skew in seconds (default 300). */
  toleranceSeconds?: number;
  /** Unix seconds "now" override (for tests); defaults to wall clock. */
  now?: number;
}

const TIMESTAMP_PATTERN = /^\d{1,15}$/;
const HEX_SIGNATURE_PATTERN = /^[0-9a-fA-F]{64}$/;

/**
 * Verify a `Coasty-Signature` header against the raw (unparsed!) request body.
 *
 * @param rawBody  The exact request bytes. Pass the raw body — re-serialized
 *                 JSON will not match.
 * @param header   The `Coasty-Signature` header value (`t=...,v1=...`).
 * @param secret   The per-run/per-trigger `webhook_secret` (shown once on create).
 * @returns `true` only when the signature is valid AND the timestamp is within
 *          tolerance. Never throws.
 */
export function verifySignature(
  rawBody: Uint8Array | string,
  header: string,
  secret: string,
  options: VerifySignatureOptions = {},
): boolean {
  try {
    if (typeof header !== 'string' || typeof secret !== 'string' || secret.length === 0) {
      return false;
    }

    const parts = new Map<string, string>();
    for (const piece of header.split(',')) {
      const separator = piece.indexOf('=');
      if (separator === -1) continue;
      const key = piece.slice(0, separator).trim();
      const value = piece.slice(separator + 1).trim();
      if (!parts.has(key)) parts.set(key, value);
    }

    const timestamp = parts.get('t');
    const signature = parts.get('v1');
    if (timestamp === undefined || signature === undefined) return false;
    if (!TIMESTAMP_PATTERN.test(timestamp)) return false;
    if (!HEX_SIGNATURE_PATTERN.test(signature)) return false;

    const tolerance = options.toleranceSeconds ?? DEFAULT_TOLERANCE_SECONDS;
    const now = options.now ?? Math.floor(Date.now() / 1000);
    if (Math.abs(now - Number(timestamp)) > tolerance) return false;

    const bodyBytes =
      typeof rawBody === 'string' ? Buffer.from(rawBody, 'utf8') : Buffer.from(rawBody);
    const signedPayload = Buffer.concat([Buffer.from(`${timestamp}.`, 'utf8'), bodyBytes]);
    const expected = createHmac('sha256', secret).update(signedPayload).digest();
    const provided = Buffer.from(signature, 'hex');

    // Equal-length guard: timingSafeEqual throws on mismatched lengths.
    if (provided.length !== expected.length) return false;
    return timingSafeEqual(expected, provided);
  } catch {
    return false;
  }
}

/**
 * Compute the `Coasty-Signature` header for a payload — used by the mock
 * server and tests to emit correctly signed webhooks.
 */
export function signPayload(
  rawBody: Uint8Array | string,
  secret: string,
  timestamp: number,
): string {
  const bodyBytes =
    typeof rawBody === 'string' ? Buffer.from(rawBody, 'utf8') : Buffer.from(rawBody);
  const signedPayload = Buffer.concat([Buffer.from(`${String(timestamp)}.`, 'utf8'), bodyBytes]);
  const signature = createHmac('sha256', secret).update(signedPayload).digest('hex');
  return `t=${String(timestamp)},v1=${signature}`;
}

File: python/src/coasty/webhooks.py
"""Coasty webhook signature verification.

Scheme (run webhooks and trigger webhooks alike):

- Header: ``Coasty-Signature: t=<unix_ts>,v1=<hex>``
- Signed payload: ``b"<t>." + raw_body``
- ``v1 = hex(HMAC_SHA256(secret, signed_payload))``

Verification uses a constant-time compare and rejects timestamps outside the
tolerance window (default +/- 300 s, the documented replay window). A
malformed header returns ``False`` -- this function never raises.
"""

from __future__ import annotations

import hashlib
import hmac
import time

DEFAULT_TOLERANCE_SECONDS = 300


def compute_signature(raw_body: bytes, secret: str, *, timestamp: int) -> str:
    """Return the hex ``v1`` signature for a body + secret + timestamp."""
    signed_payload = f"{timestamp}.".encode() + raw_body
    return hmac.new(secret.encode(), signed_payload, hashlib.sha256).hexdigest()


def build_signature_header(raw_body: bytes, secret: str, *, timestamp: int) -> str:
    """Build a ``Coasty-Signature`` header value (useful for tests/mocks)."""
    return f"t={timestamp},v1={compute_signature(raw_body, secret, timestamp=timestamp)}"


def verify_signature(
    raw_body: bytes,
    header: str,
    secret: str,
    *,
    tolerance_seconds: int = DEFAULT_TOLERANCE_SECONDS,
    now: int | None = None,
) -> bool:
    """Verify a ``Coasty-Signature`` header against the raw request body.

    Returns ``True`` only when a ``v1`` signature matches (constant-time)
    AND the ``t`` timestamp is within ``tolerance_seconds`` of ``now``
    (stale *and* future timestamps are rejected). Malformed input returns
    ``False`` -- never raises.
    """
    if not isinstance(raw_body, bytes) or not isinstance(header, str) or not secret:
        return False

    timestamp_raw: str | None = None
    candidates: list[str] = []
    for part in header.split(","):
        key, sep, value = part.strip().partition("=")
        if not sep:
            return False  # malformed part (no '=')
        key = key.strip()
        value = value.strip()
        if key == "t":
            timestamp_raw = value
        elif key == "v1":
            candidates.append(value)
        # unknown keys (e.g. future v2=...) are ignored

    if timestamp_raw is None or not candidates:
        return False
    try:
        timestamp = int(timestamp_raw)
    except ValueError:
        return False

    current = int(time.time()) if now is None else now
    if abs(current - timestamp) > tolerance_seconds:
        return False

    expected = compute_signature(raw_body, secret, timestamp=timestamp)
    return any(hmac.compare_digest(expected, candidate) for candidate in candidates)
